oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java

```diff
@@ -0,0 +1,109 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.index.indexer.document.flatfile;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.function.Predicate;
+
+import org.apache.jackrabbit.oak.index.indexer.document.NodeStateEntry;
+import org.apache.jackrabbit.oak.plugins.document.DocumentNodeState;
+import org.apache.jackrabbit.oak.spi.state.ChildNodeEntry;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeStateUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This util class can be used to export a tree (eg entire repository) to a flat
+ * file, without index dependency/involvement.
+ */
+public class SimpleFlatFileUtil {
+
+    private static final Logger log = LoggerFactory.getLogger(SimpleFlatFileUtil.class);
+
+    private static final String LINE_SEPARATOR = System.getProperty("line.separator");
+
+    private final FileWriter fw;
+    private final BufferedWriter bw;
+    private final ArrayList<StateInBytesHolder> entryBatch = new ArrayList<>();
+    private final Predicate<String> pathPredicate = path -> true;
+    private final NodeStateEntryWriter entryWriter;
+    private long totalLines = 0;
+
+    private SimpleFlatFileUtil(File f) throws IOException {
+        // blobStore is only used for deserialization - so pass null here:
+        entryWriter = new NodeStateEntryWriter(null);
+        fw = new FileWriter(f);
+        bw = new BufferedWriter(fw);
+    }
+
+    public static void createFlatFileFor(NodeState ns, File f) throws IOException {
+        final SimpleFlatFileUtil h = new SimpleFlatFileUtil(f);
+        log.info("createFlatFileFor : writing to {}", f.getCanonicalPath());
+        h.addEntryAndTraverseChildren(ns);
+        h.close();
+        log.info("createFlatFileFor : done. wrote {} lines in total.", h.totalLines);
+    }
+
+    private void close() throws IOException {
+        flush();
+        bw.close();
+        fw.close();
+    }
+
+    private void flush() throws IOException {
+        for (StateInBytesHolder nsh : entryBatch) {
+            String line = entryWriter.toString(nsh.getPathElements(), nsh.getLine());
+            bw.append(line);
+            bw.append(LINE_SEPARATOR);
+        }
+        log.info("flush : wrote another {} nodes, total so far: {} lines.",
+                entryBatch.size(), totalLines);
+        totalLines += entryBatch.size();
+        entryBatch.clear();
+    }
+
+    private void addEntryAndTraverseChildren(NodeState ns) throws IOException {
+        addEntry(ns);
+        if (entryBatch.size() > 999) {
+            flush();
+        }
+        for (ChildNodeEntry e : ns.getChildNodeEntries()) {
```

Comment:
If I understood correctly, this will be querying directly the local MongoDB instance. And this is doing a depth-first traversal of the tree rooted in NodeState given to the createFlatFileFor method. Is return of `getChildNodeEntries` deterministic? That is, can you be sure that the FFS file will always be the same for repositories with equivalent contents? Running a revision GC might change the order by which Mongo answers to some queries, even if the results are equivalent. If the order of results of `getChildNodeEntries` is not deterministic, then you could sort the child nodes before iterating over them.

Reply:
Good point, I agree this should be addressed. Actually there's also the issue that `NodeStateEntryWriter` excludes the `:childOrder`.

I'm not sure what the best approach is yet. Might have to do some iteration over this code..

Reply:
PS: also noticed that `nodeState.getProperties()` does not always return the same order - so the properties probably need to be sorted explicitly (in the `NodeStateEntryWriter`)

Reply:
I have now added sorting of both children and properties (to the `SimpleFlatFileUtil` only, not by default to underlying classes). Also added an option to export the `:childOrder` property - this allows explicit veriication of this property (given the sorting does alphabetic sorting now anyway)